### ISSUE TITLE
tracking issue for dtype refactor

### ISSUE THREAD

**Issue by chenyuxyz** [tinygrad/tinygrad#2721]
- [x] type promotion helpers (List[input_type] -> output_type) #2724 
- [x] update and fix test_uops where to have vin[0] as bool #2782
- [x] int / int list should be treated as int32 tensor #2789
- [x] tensor.py handles unary that need to cast to float (exp, sin... etc) #2740 #2887 
- [x] mlops handles least upper type for elementwise ops #2788 
- [x] dtypes.int+3.3 -> float, dtypes.half+3.3 -> half #2785 
- [x] Tensor of list of list of list... of int should be dtypes.int #2793
- [x] Tensor.default_type -> dtypes.default_float #2824
- [x] Tensor.sum returns at least int32 or float32 #2926 
- [x] all Tensor cmp returns in dtypes.bool #2931 
- [x] assert that type is properly handled in uops level, and assert in lazy level #2779 #2782 #2800 #2888
- [x] remove match_types in ops_cpu and ops_torch once all input types are aligned in ops #2817
- [x] remove output_type in ops_cpu and ops_torch once all output types are aligned in ops #2892
- [x] skipped onnx sce test in #2785, reenabled in #2902
- [x] llama is quite slower, fix type for freq and mask #2932
- [x] confirm argmax and argmin returns dtypes.int see #2818 #2933
- [ ] ~~remove the hack in WEBGPU to write bool as f32~~
- [ ] cannot create a bfloat16 tensor. see #2795
- [x] remove the redundant manual cast and manual dtype in tensor.py and onnx #2939

i started to write some of it in (https://github.com/chenyuxyz/tinygrad/pull/28) and realized lack of helper to get the output type is blocking everything, so i will do that next. @geohotstan @Qazalin fyi.

for the type promotion we will follow the jax rules. and we will try to match jax behavior (already found some undesired behavior in torch...)
https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.dtype.html
https://github.com/google/jax/blob/3651d4c4f5d59c570f573824e32c0515a8c883f0/jax/_src/dtypes.py#L556
https://github.com/google/jax/blob/b225c86f10380ee7cb924153b6349ef1359f7a15/tests/dtypes_test.py#L126
